servant-client/src/Servant/Client/Class.hs

```diff
@@ -0,0 +1,16 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-| Types for possible backends to run client-side `Req` queries -}
+module Servant.Client.Class
+  (RunClient(..))
+where
+
+import Data.Proxy
+import Network.HTTP.Types
+import Servant.API
+import Servant.Common.Req
+
+class (Monad m) => RunClient m ct result where
+  runRequest :: Proxy ct
```

Comment:
@jkarni, I don't think `Free` is good here. From what I look, it would be enough to have `Functor` constraint here, not a `Monad`. Could @abailly try if it's enough,

```diff
-  clientWithRoute pm Proxy req = do
-    (_hdrs, a) :: ([H.Header], a) <- runRequest (Proxy :: Proxy ct) method req
-    return a
+  clientWithRoute pm Proxy req =
+    fmap snd (runRequest (Proxy :: Proxy ct) method req :: m ([H.Header], a))
```

etc.

Reply:
Tried it but it does not seem to work: I need to _sequence_ evaluation before return value in the `NoContent` case